// Modules generated by transparent proc macros still acts as barriers for names (issue #50504).

// aux-build:generate-mod.rs

extern crate generate_mod;

struct FromOutside;

generate_mod::check!(); // { dg-error ".E0412." "" { target *-*-* } }
// { dg-error ".E0412." "" { target *-*-* } .-2 }

#[generate_mod::check_attr] // { dg-error ".E0412." "" { target *-*-* } }
// { dg-error ".E0412." "" { target *-*-* } .-2 }
struct S;

#[derive(generate_mod::CheckDerive)] // { dg-warning ".E0412." "" { target *-*-* } }
// { dg-warning "" "" { target *-*-* } .-2 }
// { dg-warning "" "" { target *-*-* } .-3 }
// { dg-warning "" "" { target *-*-* } .-4 }
struct Z;

fn inner_block() {
    #[derive(generate_mod::CheckDerive)] // { dg-warning ".E0412." "" { target *-*-* } }
// { dg-warning "" "" { target *-*-* } .-2 }
// { dg-warning "" "" { target *-*-* } .-3 }
// { dg-warning "" "" { target *-*-* } .-4 }
    struct InnerZ;
}

#[derive(generate_mod::CheckDeriveLint)] // OK, lint is suppressed
struct W;

fn main() {}

