/* { dg-do preprocess } */
/* { dg-options "-std=c11 -pedantic" { target c } } */
/* { dg-options "-std=c++11 -pedantic" { target c++ } } */

¨

B̀

̀ /* { dg-error "not valid at the start of an identifier" } */

À /* { dg-warning "not in NFC" } */

𐀀
🿽
󡈴
