/* { dg-do preprocess } */
/* { dg-options "-std=gnu++98 -pedantic" } */

ª /* { dg-error "not valid in an identifier" } */
« /* { dg-error "not valid in an identifier" } */
¶ /* { dg-error "not valid in an identifier" } */
º /* { dg-error "not valid in an identifier" } */
À
Ö
΄

٩ /* { dg-error "not valid in an identifier" } */
A٩ /* { dg-error "not valid in an identifier" } */
0º /* { dg-error "not valid in an identifier" } */
0٩ /* { dg-error "not valid in an identifier" } */
๙
A๙
