pub fn main() {
    while let  = 5
    // { dg-error "should be at least 1 pattern" "" { target *-*-* } .-1 }
    // { dg-error "failed to parse statement or expression in block expression" "" { target *-*-* } .-2 }
    // { dg-error "unrecognised token .=. for start of item" "" { target *-*-* } .-3 }
    // { dg-error "failed to parse item in crate" "" { target *-*-* } .-4 }
    {}
}

