fn main() {}

fn v() -> isize { // { dg-error ".E0308." "" { target *-*-* } }
mod _ { // { dg-error "" "" { target *-*-* } }
pub fn g() -> isizee { // { dg-error ".E0412." "" { target *-*-* } }
mod _ { // { dg-error "" "" { target *-*-* } }
pub    g() -> is // { dg-error "" "" { target *-*-* } }
(), w20);
}
(), w20); // { dg-error "" "" { target *-*-* } }
}

