fn main() {
    let _: usize = ()
// { dg-error ".E0308." "" { target *-*-* } .-1 }
// { dg-error ".E0308." "" { target *-*-* } .-2 }
    let _ = 3;
}

fn foo() -> usize {
    let _: usize = ()
// { dg-error ".E0308." "" { target *-*-* } .-1 }
// { dg-error ".E0308." "" { target *-*-* } .-2 }
    return 3;
}

