fn main() {
    let 1234567890123456789012345678901234567890e-340: f64 = 0.0;
// { dg-error ".E0080." "" { target *-*-* } .-1 }

    fn param(1234567890123456789012345678901234567890e-340: f64) {}
// { dg-error ".E0080." "" { target *-*-* } .-1 }
}

